File: ai-backend/src/services/data_source/data_source_config_schema.py
from typing import Optional

from pydantic import BaseModel, Field, field_validator


def trim_and_validate_space(v):
    v = v.strip() if v else None
    if v and " " in v:
        raise ValueError("must not contain spaces")
    return v


# Postgres Config
class PostgresConfig(BaseModel):
    host: str = Field(..., description="Database host")
    port: int = Field(default=5432, description="Database port")
    database: str = Field(..., description="Database name")
    username: str = Field(..., description="Database username")
    password: str = Field(..., description="Database password")
    ssl_mode: Optional[str] = Field(default="require", description="SSL mode")
    connect_timeout: Optional[int] = Field(default=10, description="Connection timeout in seconds")

    @field_validator("username", "database", "host")
    def trim_and_validate_space(cls, v):
        return trim_and_validate_space(v)


# MySQL Config
class MySQLConfig(BaseModel):
    host: str = Field(..., description="Database host")
    port: int = Field(default=3306, description="Database port")
    database: str = Field(..., description="Database name")
    username: str = Field(..., description="Database username")
    password: str = Field(..., description="Database password")
    ssl_ca: Optional[str] = Field(default=None, description="SSL CA certificate path")
    connect_timeout: Optional[int] = Field(default=10, description="Connection timeout in seconds")

    @field_validator("username", "database", "host")
    def trim_and_validate_space(cls, v):
        return trim_and_validate_space(v)


# Snowflake Config
class SnowflakeConfig(BaseModel):
    account_identifier: str = Field(..., description="Snowflake account identifier")
    warehouse: str = Field(..., description="Warehouse name")
    database: str = Field(..., description="Database name")
    username: str = Field(..., description="Username")
    password: str = Field(..., description="Password")
    role: Optional[str] = Field(default=None, description="Role name")
    schema: Optional[str] = Field(default=None, description="Schema name")

    @field_validator("account_identifier", "warehouse", "database", "username")
    def trim_and_validate_space(cls, v):
        return trim_and_validate_space(v)


class SQLiteConfig(BaseModel):
    database_path: str = Field(..., description="SQLite database path. Can only be used by ")

File: analyst-ui/services/dataSource/dataSourceConfig.schema.ts
import { z } from "zod";

// Helper function to trim and validate space
const trimAndValidateSpace = (value: string) => {
  const trimmed = value.trim();
  if (trimmed.includes(" ")) {
    throw new Error("Value cannot contain spaces");
  }
  return trimmed;
};

// Postgres Config
export const PostgresConfigSchema = z.object({
  host: z.string().transform(trimAndValidateSpace),
  port: z.coerce.number(),
  database: z.string().transform(trimAndValidateSpace),
  username: z.string().transform(trimAndValidateSpace),
  password: z.string().transform(trimAndValidateSpace),
  ssl_mode: z.string().default("require").optional(),
  connect_timeout: z.number().default(10).optional(),
});

// MySQL Config
export const MySQLConfigSchema = z.object({
  host: z.string().transform(trimAndValidateSpace),
  port: z.coerce.number(),
  database: z.string().transform(trimAndValidateSpace),
  username: z.string().transform(trimAndValidateSpace),
  password: z.string().transform(trimAndValidateSpace),
  ssl_ca: z.string().optional(),
  connect_timeout: z.number().default(10).optional(),
});

// Snowflake Config
export const SnowflakeConfigSchema = z.object({
  account_identifier: z.string().transform(trimAndValidateSpace),
  username: z.string().transform(trimAndValidateSpace),
  password: z.string().transform(trimAndValidateSpace),
  database: z.string().transform(trimAndValidateSpace),
  warehouse: z.string().transform(trimAndValidateSpace),
  schema: z.string().transform(trimAndValidateSpace).optional().nullable(),
  role: z.string().optional().nullable(),
});

// SQLite Config
export const SQLiteConfigSchema = z.object({
  database_path: z.string().transform(trimAndValidateSpace),
});
